testing/integration/kubernetes_agent_standalone_test.go

```diff
@@ -937,43 +861,260 @@ func generateESAPIKey(esClient *elasticsearch.Client, keyName string) (string, e
+			case *rbacv1.ClusterRoleBinding:
+				for idx := range objWithType.Subjects {
+					objWithType.Subjects[idx].Namespace = opts.namespace
+				}
+				continue
+			case *rbacv1.RoleBinding:
+				for idx := range objWithType.Subjects {
+					objWithType.Subjects[idx].Namespace = opts.namespace
+				}
+				continue
+			}
+		}
+		if err := client.Resources().Create(ctx, obj); err != nil {
+			return fmt.Errorf("failed to create object %s: %w", obj.GetName(), err)
+		}
+	}
+
+	if !opts.wait {
+		// no need to wait
+		return nil
+	}
+
+	if opts.waitTimeout == 0 {
+		// default to 120 seconds
+		opts.waitTimeout = 120 * time.Second
+	}
+
+	return k8sWaitForReady(ctx, client, opts.waitTimeout, objects...)
+}
+
+// k8sWaitForReady waits for the given k8s objects to be ready
+func k8sWaitForReady(ctx context.Context, client klient.Client, waitDuration time.Duration, objects ...k8s.Object) error {
+	// use ready checker from helm kube
+	clientSet, err := kubernetes.NewForConfig(client.RESTConfig())
+	if err != nil {
+		return fmt.Errorf("error creating clientset: %w", err)
+	}
+	readyChecker := helmKube.NewReadyChecker(clientSet, func(s string, i ...interface{}) {})
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, waitDuration)
+	defer cancel()
+
+	waitFn := func(ri *cliResource.Info) error {
+		// here we wait for the k8s object (e.g. deployment, daemonset, pod) to be ready
+		for {
+			ready, readyErr := readyChecker.IsReady(ctxTimeout, ri)
+			if ready {
+				// k8s object is ready
+				return nil
+			}
+			// k8s object is not ready yet
+			readyErr = errors.Join(fmt.Errorf("k8s object %s is not ready", ri.Name), readyErr)
+
+			if ctxTimeout.Err() != nil {
+				// timeout
+				return errors.Join(fmt.Errorf("timeout waiting for k8s object %s to be ready", ri.Name), readyErr)
+			}
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+
+	for _, o := range objects {
+		// convert k8s.Object to resource.Info for ready checker
+		runtimeObj, ok := o.(runtime.Object)
+		if !ok {
+			return fmt.Errorf("unable to convert k8s.Object %s to runtime.Object", o.GetName())
+		}
+
+		if err := waitFn(&cliResource.Info{
+			Object:    runtimeObj,
+			Name:      o.GetName(),
+			Namespace: o.GetNamespace(),
+		}); err != nil {
+			return err
+		}
+		// extract pod label selector for all k8s objects that have underlying pods
+		oPodsLabelSelector, err := helmKube.SelectorsForObject(runtimeObj)
+		if err != nil {
+			// k8s object does not have pods
+			continue
+		}
+
+		podList, err := clientSet.CoreV1().Pods(o.GetNamespace()).List(ctx, metav1.ListOptions{
+			LabelSelector: oPodsLabelSelector.String(),
+		})
+		if err != nil {
+			return fmt.Errorf("error listing pods: %w", err)
+		}
+
+		// here we wait for the all pods to be ready
+		for _, pod := range podList.Items {
+			if err := waitFn(&cliResource.Info{
+				Object:    &pod,
+				Name:      pod.Name,
+				Namespace: pod.Namespace,
+			}); err != nil {
+				return err
+			}
+		}
```

Comment:
Do we actually need to do this? It looks like the [implementation](https://github.com/helm/helm/blob/b91a772d71340fcc250bd0616c8b6f596dca5746/pkg/kube/ready.go#L92) of the ready checker already checks Pods for resources where that's appropriate.

Reply:
I like the way you think @swiatekm 🙂  but since you are looking at the code you missed out the fact that Daemonsets/Statefulset/Deployment have default strategies that affect the ready checks, e.g. for Daemonset there is the MaxUnavailable option which by default is 1 and you run on a single node cluster, as we do with kind, this can lead to false positives in a way ([code ref](https://github.com/helm/helm/blob/b91a772d71340fcc250bd0616c8b6f596dca5746/pkg/kube/ready.go#L326)). That why I went the extra mile, of checking all associated pods. With that explanation does it seem appropriate? 🙂